feat: API spec for Revocations Check endpoint (#141)
web3-storage/specs: w3-revocations-check.md

# Capabilities

## Revocation Check

### Endpoint Details

- **Method:** `GET`
- **Path:** `/{cid}` (can be hosted at any path, the CID parameter is the key component)
- **Authentication:** None required (public endpoint)
- **CDN Cacheable:** Yes (simple GET requests with 200/404 responses)

### Request Format

#### URL Parameters

**`cid`** (required): Delegation CID to check for revocations. Must be a valid IPFS CID string.

#### Constraints

- **CID Format:** Valid IPFS CID string (e.g., `bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi`)
- **Encoding:** the CID SHOULD be encoded using multibase base32 (e.g., `bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi`)

### Response Format

#### Success Response (200): Delegation Revoked

**Content-Type:** `application/vnd.ipld.car`

Returns a CAR (Content Addressable aRchive) file containing:

- Revocation metadata in `revocations@0.0.1` schema
- Embedded UCAN proof blocks for trustless verification
- ETag header for caching optimization

##### CAR File Structure

> The CAR file contains a root block with the following structure (shown in DAG-JSON for readability, but it is **RECOMMENDED** that the root block be encoded as [DAG-CBOR](https://ipld.io/docs/codecs/known/dag-cbor/) for efficiency):

Reviewer: Capitals is enough 🤣 

```suggestion
The CAR file contains a root block with the following structure (shown in DAG-JSON for readability, but it is RECOMMENDED that the root block be encoded as [DAG-CBOR](https://ipld.io/docs/codecs/known/dag-cbor/) for efficiency):
```

```json
{
  "revocations@0.0.1": {
    "revocations": [
      {
        "delegation": { "/": "[delegationCID]" },

>         "scope": "[scopeDID]", 

Reviewer: ```suggestion
```

        "cause": { "/": "[causeCID]" }
      }
    ]
  }
}
```

**Field Descriptions:**

- **`delegation`**: CID of the delegation that has been revoked

> - **`scope`**: DID of the authority that issued the revocation (either the issuer or audience of the delegation under inspection, or one of its proofs). *Note: This field is derivable from the cause invocation but included for convenience*

Reviewer: What is the convenience? I just...can't think of a reason.

```suggestion
```

## Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119).

# Introduction

The revocations check protocol allows clients to verify whether UCAN delegations have been explicitly revoked. The protocol provides a simple HTTP GET endpoint that returns either cryptographic proof of revocation or indicates that no revocation exists for the queried delegation CID.

# Capabilities

## Revocation Check

### Endpoint Details

- **Method:** `GET`
- **Path:** `/revocations/{cid}`
- **Authentication:** None required (public endpoint)
- **CDN Cacheable:** Yes (simple GET requests with 200/404 responses)

### Request Format

#### URL Parameters

**`cid`** (required): Delegation CID to check for revocations. Must be a valid IPFS CID string.

#### Constraints

- **CID Format:** Valid IPFS CID string (e.g., `bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi`)

> - **Encoding:** CID will be normalized to string format internally

Reviewer: ```suggestion
- **Encoding:** the CID SHOULD be encoded using multibase base32
```

### Response Format

#### Success Response (200): Delegation Revoked

**Content-Type:** `application/vnd.ipld.car`

Returns a CAR (Content Addressable aRchive) file containing:

- Revocation metadata in `revocations@0.0.1` schema
- Embedded UCAN proof blocks for trustless verification
- ETag header for caching optimization

##### CAR File Structure

The CAR file contains a root block with the following structure:

```json
{
  "revocations@0.0.1": {
    "revocations": [
      {
        "delegation": { "/": "[delegationCID]" },

>         "scope": "[scopeDID]", 

Reviewer: I don't think we need to include this, as it is derivable from the cause invocation.

Note: scope is the DID of the authority that issued the revocation - either the issuer or audience of the delegation that is under inspection or one of it's proofs.

### Error Responses

#### 400 Bad Request

**Content-Type:** `application/json`

##### Invalid CID Parameter

```json
{
  "error": "Bad request",
  "message": "Invalid CID parameter"
}
```

##### Missing CID Parameter

```json
{
  "error": "Bad request", 
  "message": "CID parameter is required"
}
```

#### 500 Internal Server Error

**Content-Type:** `application/json`

##### Query Failure

```json
{
  "error": "Internal server error",
  "message": "Failed to query revocations"
}
```

##### General Error

```json
{
  "error": "Internal server error",
  "message": "An unexpected error occurred"
}
```

### CORS Headers

The endpoint includes CORS headers for cross-origin requests:

- `Access-Control-Allow-Origin: *`
- `Access-Control-Allow-Methods: GET`
- `Access-Control-Allow-Headers: Accept`

# Usage Examples

## cURL Example

```bash
# Check if a delegation is revoked
curl -X GET https://up.storacha.network/revocations/{cid}

# Save CAR file for offline verification (when response is 200)
curl -X GET https://up.storacha.network/revocations/{cid} \
  -H "Accept: application/vnd.ipld.car" \
  -o revocation-proof.car
```

## Client-Side Proof Chain Verification

```javascript
/**

>  * Recursively verify the delegation proof chain for revocations

Reviewer: I would not...recurse. I'd collect the CIDs we want to check (start of chain first) and make parallel requests (with concurrency limit) allowing cancellation if one is found to be revoked.

I don't think it's worth putting this example in the spec - I'd probably remove.

 */

> async function verifyProofChain(delegation) {

Reviewer: This is more `isRevoked(delegation)` or `checkRevocation(delegation)`.

I'd maybe return `Result<Unit, Revoked | Failure>`, where `Revoked` can carry information about the delegation that was revoked and the cause.

```ts
interface Revoked extends Failure {
  name: 'Revoked'
  /** The delegation/proof that was revoked */
  delegation: Delegation
  /** The invocation that revoked the delegation */
  cause: Invocation
}
```

  // Check if delegation is explicitly revoked
  const response = await fetch(`https://up.storacha.network/revocations/${delegation.cid}`);
  
  if (response.status === 200) {

>     return { isValid: false, reason: 'Delegation explicitly revoked' };

Reviewer: Note: this is not doing verification, it's _trusting_ the 200 from the service.

Would be worth building a client library that does the verification.

I would perhaps check from the start of the delegation chain first.
